In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

# Q1. Pandas version

In [2]:
pd.__version__

'2.2.2'

**********

# Q2. Records count \
How many records are in the dataset?

1) 12
2) 1000
3) 2160
4) 12160

# Answer
3) 2160

In [3]:
dataframe = pd.read_csv('data/laptops.csv')
dataframe

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

***

# Q3. Laptop brands \
How many laptop brands are presented in the dataset?

1) 12
2) 27
3) 28
4) 2160

# Answer
2) 27

In [4]:
unique_laptop_brands = dataframe['Brand'].unique()
print(f'Total Unique Laptop Brands = {unique_laptop_brands}')
print('****************************************************************')
print(f'Total Number of unique Laptop Brands = {len(unique_laptop_brands)}')

Total Unique Laptop Brands = ['Asus' 'Alurin' 'MSI' 'HP' 'Lenovo' 'Medion' 'Acer' 'Apple' 'Razer'
 'Gigabyte' 'Dell' 'LG' 'Samsung' 'PcCom' 'Microsoft' 'Primux' 'Prixton'
 'Dynabook Toshiba' 'Thomson' 'Denver' 'Deep Gaming' 'Vant' 'Innjoo'
 'Jetwing' 'Millenium' 'Realme' 'Toshiba']
****************************************************************
Total Number of unique Laptop Brands = 27


***

# Q4. Missing values\
How many columns in the dataset have missing values?

1) 0
2) 1
3) 2
4) 3

# Answer
4) 3

In [5]:
# missing_values = dataframe.isnull().sum()
columns_missing_values = dataframe.isnull().sum() > 0
number_columns_with_missing = columns_missing_values.sum()

print(f'Total number of columns with missing values = {number_columns_with_missing}')

Total number of columns with missing values = 3


****

# Q5. Maximum final price\
What's the maximum final price of Dell notebooks in the dataset?

1) 869
2) 3691
3) 3849
4) 3936

# Answer
4) 3936

In [6]:
# dell_notebook_max_price = dataframe[(dataframe['Brand'] == 'Dell')]
dell_notebook_max_price = dataframe[(dataframe['Brand'] == 'Dell')]['Final Price'].max()
print(f'Maximum Final Price of Dell Notebook = {int(dell_notebook_max_price)}')

Maximum Final Price of Dell Notebook = 3936


*****

# Q6. Median value of Screen

## Task
### Task 1) Find the median value of Screen column in the dataset.
### Task 2) Next, calculate the most frequent value of the same Screen column.
### Task 3) Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
### Task 4) Now, calculate the median value of Screen once again.

In [7]:
# Task-01
median_value_screen = dataframe['Screen'].median()
median_value_screen

np.float64(15.6)

In [8]:
# Task-02
mode_frequent_value = dataframe['Screen'].mode()[0]
mode_frequent_value

np.float64(15.6)

In [9]:
# Task-03
dataframe['Screen'].fillna(mode_frequent_value, inplace=True)
# dataframe

In [10]:
# Task - 04
final_median = dataframe['Screen'].median()
final_median

np.float64(15.6)

# Answer for Question 6 is 'NO', Median value did not change

************************************

# Q7. Sum of weights

#### Task 1) Select all the "Innjoo" laptops from the dataset.

In [11]:
injOO_laptops = dataframe[dataframe['Brand'] == 'Innjoo']
injOO_laptops

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

In [12]:
#### Task 2) Select only columns RAM, Storage, Screen.
injoo_ram_storage_screen = injOO_laptops[['RAM', 'Storage', 'Screen']]
injoo_ram_storage_screen

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [13]:
#### Task 3) Get the underlying NumPy array. Let's call it X.
X = injoo_ram_storage_screen.to_numpy()
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [14]:
#### Task 4) Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

XTX = X.T @ X
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [15]:
#### Task 5) Compute the inverse of XTX.
XTX_inverse = np.linalg.inv(XTX)
XTX_inverse

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [16]:
#### Task 6) Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
y = np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])

In [17]:
#### Task 7) Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = XTX_inverse @ X.T @ y
w

array([45.58076606,  0.42783519, 45.29127938])

In [19]:
#### Task 8) What's the sum of all the elements of the result?
sum_of_elements = np.sum(w)
print(f'Sum of all the elements (Sum of weights) = {sum_of_elements}')

Sum of all the elements (Sum of weights) = 91.2998806299555


##### Answer for Question 7 is 91.30

# End of Week 01 Home Work